staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/listtype/validation.go

```diff
@@ -0,0 +1,228 @@
+	case map[string]interface{}:
+		return ValidateListSetsAndMaps(fldPath, s, obj)
+	}
+	return nil
+}
+
+func validateListSetsAndMapsArray(fldPath *field.Path, s *schema.Structural, obj []interface{}) field.ErrorList {
+	var errs field.ErrorList
+
+	if s.XListType != nil {
+		switch *s.XListType {
+		case "set":
+			dup, err := validateListSet(fldPath, obj)
+			if err != nil {
+				errs = append(errs, err)
+			} else if dup >= 0 {
+				errs = append(errs, field.Duplicate(fldPath.Index(dup), obj[dup]))
+			}
+		case "map":
+			errs = append(errs, validateListMap(fldPath, s, obj)...)
+		}
+	}
+
+	if s.Items != nil {
+		for i := range obj {
+			errs = append(errs, validationListSetAndMaps(fldPath.Index(i), s.Items, obj[i])...)
+		}
+	}
+
+	return errs
+}
+
+func validateListSet(fldPath *field.Path, obj []interface{}) (int, *field.Error) {
+	if len(obj) <= 1 {
+		return -1, nil
+	}
+	if len(obj) == 2 {
+		if reflect.DeepEqual(obj[0], obj[1]) {
+			return 1, nil
+		}
+	}
+
+	seenScalars := make(map[interface{}]bool, len(obj))
+	seenCompounds := make(map[string]bool, len(obj))
+	for i, x := range obj {
+		switch x.(type) {
+		case map[string]interface{}, []interface{}:
+			bs, err := json.Marshal(x)
+			if err != nil {
+				return -1, field.Invalid(fldPath.Index(i), x, "internal error")
+			}
+			s := string(bs)
+			if seenCompounds[s] {
+				return i, nil
+			}
+			seenCompounds[s] = true
+		default:
+			if seenScalars[x] {
+				return i, nil
+			}
+			seenScalars[x] = true
+		}
+	}
+
+	return -1, nil
+}
+
+func validateListMap(fldPath *field.Path, s *schema.Structural, obj []interface{}) field.ErrorList {
+	// only allow nil and objects, find nil duplicates
+	seenNil := false
+	for i, x := range obj {
+		if x == nil {
+			if seenNil {
+				return field.ErrorList{field.Duplicate(fldPath.Index(i), x)}
+			}
+			seenNil = true
+		} else if _, ok := x.(map[string]interface{}); !ok {
+			return field.ErrorList{field.Invalid(fldPath.Index(i), x, "must be an object for an array of list-type map")}
+		}
+	}
+
+	if len(obj) <= 1 {
+		return nil
+	}
+
+	// optimize simple case of one key
+	if len(s.XListMapKeys) == 1 {
+		keyField := s.XListMapKeys[0]
+		keys := make([]interface{}, 0, len(obj))
+		seenUndefinedKey := false
+		for i, x := range obj {
+			if x == nil {
+				continue
+			}
+
+			x := x.(map[string]interface{})
+
+			// undefined key?
+			key, ok := x[keyField]
+			if !ok {
+				if seenUndefinedKey {
+					return field.ErrorList{field.Duplicate(fldPath.Index(i), x)}
+				}
+				seenUndefinedKey = true
+				continue
+			}
+
+			keys = append(keys, key)
+		}
+
+		dup, err := validateListSet(fldPath, keys)
+		if err != nil {
+			return field.ErrorList{err}
+		} else if dup >= 0 {
+			// find dup'th item, skipping nil and with undefined key
+			i := 0
+			n := dup
+			for {
+				for ; obj[i] == nil || (obj[i]).(map[string]interface{})[keyField] == 0; i++ {
+				}
+				if n == 0 {
+					break
+				}
+				i++
+				n--
+			}
```

Comment:
if we treat nil/no-key items consistently by appending them, I think we can drop this

Reply:
Fixed with `struct{}` aliases.